aiohttp/web_jobs.py

```diff
@@ -0,0 +1,223 @@
+
+    def _start(self):
+        assert not self._closed
+        self._task = ensure_future(self._coro, loop=self._loop)
+        self._task.add_done_callback(self._done_callback)
+
+    def __repr__(self):
+        info = []
+        if self._closed:
+            info.append('closed')
+        elif self._task is None:
+            info.append('pending')
+        info = ' '.join(info)
+        if info:
+            info += ' '
+        return '<Job {}coro=<{}>>'.format(info, self._coro)
+
+    @asyncio.coroutine
+    def wait(self, timeout=None):
+        self._explicit_wait = True
+        try:
+            with async_timeout.timeout(timeout=timeout, loop=self._loop):
+                return (yield from self._task)
+        except asyncio.TimeoutError as exc:
+            yield from self.close()
+            raise exc
+
+    @property
+    def closed(self):
+        return self._closed
+
+    @asyncio.coroutine
+    def close(self):
+        if self._task is None:
+            self._closed = True
+            self._runner._done(self, True)
+            return
+        if self._closed:
+            return
+        self._task.cancel()
+        # self._runner is None after _done_callback()
+        runner = self._runner
+        try:
+            with async_timeout.timeout(timeout=self._runner._timeout,
+                                       loop=self._loop):
+                yield from self._task
+        except asyncio.CancelledError:
+            pass
+        except asyncio.TimeoutError as exc:
+            context = {'message': "Job closing timed out",
+                       'job': self,
+                       'exception': exc}
+            if self._source_traceback is not None:
+                context['source_traceback'] = self._source_traceback
+            runner.call_exception_handler(context)
+
+    def _done_callback(self, task):
+        runner = self._runner
+        runner._done(self, False)
+        try:
+            exc = task.exception()
+        except asyncio.CancelledError:
+            pass
+        else:
+            if exc is not None and not self._explicit_wait:
+                context = {'message': "Job processing failed",
+                           'job': self,
+                           'exception': exc}
+                if self._source_traceback is not None:
+                    context['source_traceback'] = self._source_traceback
+                runner.call_exception_handler(context)
+                runner._failed_tasks.put_nowait(task)
+        self._runner = None  # drop backref
+        self._closed = True
+
+
+class JobRunner(Container):
+    def __init__(self, *, loop):
+        self._loop = loop
+        self._jobs = set()
+        self._timeout = 0.1
+        self._concurrency = 100
+        self._exception_handler = None
+        self._failed_tasks = asyncio.Queue(loop=loop)
+        self._failed_waiter = ensure_future(self._wait_failed(), loop=loop)
+        self._pending = deque()
```

Comment:
I think pending list should be exposed as well. This class is basically queue of tasks which may overflow and you would like to catch that moment until it will be too late.